### PR TITLE
Textures and Refactoring

### PR DESCRIPTION
**Pull request by MarcasRealAccount**
## Textures
I added both `stb_image` and `stb_truetype`, since we probably will have text and so why not use stb's truetype for that. (I just wanna clarify that they are in their own project as well)
Currently I've implemented
- Texture2D
- Texture2DArray
- Texture3D
- TextureCubeMap
- Loading of Texture2D's

## Refactoring
I've gone through every file and added a few namespaces to them to organize the classes and structs, such that not everything pops up when doing `"gp1::" + Shift Space`, just to clarify their namespaces are very similar to their actual file path.
I've removed the `gp1` namespace from every game file, since we don't need them there.
I've also gone through every file and pressed `Shift F` which causes a format document to happen so that every file uses the same format.
I've also fixed a few bugs that I can't remember at the top of my head.
I also made most variables and functions follow the same format.
I also added short comments to some if not all functions (most likely not all).
I made some small changes to how the Input System handles axis events such that it also transmits the previous value of the axis or the current if none was found, this is to make it feel more correct when wanting good delta when changing between two `InputGroup`'s, there's still a small issue when holding down a button and changing `InputGroup`, due to the release button not being transmitted, so that needs to fixed at some point.
I put the `OpenAL` and `dr_libs` stuff their own projects. (Btw to make it be implemented in their projects I had to make a few files in the `third_party` folder itself
Also I made the project compile with no warnings from **MSVC**

## Freecam
I also took the freecam and integrated it into the `Camera` class, though I had to do some small adjustments to make it work in a mostly good way.
To use freecam you will have to press the window

## Note
Yes I fully know and would agree that this pull request is just insane but it's mostly because I forgot to do the refactoring on it's own so I just continued to do it, but if you want me to extract the texture stuff from this one I can fully do that.
Though that would obviously make it take a bit longer before textures are completed.

### CODE REVIEW

**Review by sfulham — APPROVED** [review]
Looks good to me!

**Review by Gaztin — CHANGES_REQUESTED** [review]
I think we should keep getters as one-liners. The header files look a little busy to me!

File: src/Engine/Application.cpp [hunk 1]
```diff
@@ -36,57 +44,141 @@ namespace gp1 {
 		m_Mesh->m_Indices.push_back(2);
 		m_Mesh->m_Indices.push_back(1);
 
-		m_Material->SetShader(Shader::GetShader("shader"));
+		m_Material->SetShader(renderer::shader::Shader::GetShader("shader"));
```
Comment (Gaztin) [inline src/Engine/Application.cpp:47]: Should maybe rename the shader to "shader_mesh_default" or something along those lines

File: src/Engine/Renderer/Apis/OpenGL/Shader/OpenGLMaterialData.cpp [hunk 2]
```diff
@@ -1,168 +1,304 @@
...
-		for (auto& uniform : GetUniforms()) {
-			const std::string& id = uniform.first;
-			uint32_t location = GetMaterial<Material>()->GetShader()->GetUniformLocation(id);
-			if (uniform.second.type() == typeid(Uniform<float>)) {
-				const Uniform<float>* uniformFloat = std::any_cast<Uniform<float>>(&uniform.second);
+			const std::string& id       = uniform.first;
+			uint32_t           location = material->GetShader()->GetUniformLocation(id);
+			if (uniform.second.type() == typeid(Uniform<float>))
+			{
+				const Uniform<float>* uniformFloat = uniform.second._Cast<Uniform<float>>();
```
Comment (Gaztin) [inline src/Engine/Renderer/Apis/OpenGL/Shader/OpenGLMaterialData.cpp:63]: What is `_Cast`? If that is an MSVC-specific function then it's a big no-no.

Comment (MarcasRealAccount) [reply src/Engine/Renderer/Apis/OpenGL/Shader/OpenGLMaterialData.cpp:63]: Isn't it weird how only MSVC has functions that make it look nicer.
But yeah I just tested and the `_Cast` is sadly an MSVC-specific function, so I'll get to swapping them back to `std::any_cast`

File: src/Engine/Application.cpp (hunk 1 above)
Comment (MarcasRealAccount) [reply src/Engine/Application.cpp:47]: Sure, though it was only supposed to be a test shader.

**Comment by MarcasRealAccount** [comment]
> I think we should keep getters as one-liners. The header files look a little busy to me!

Which getters are you referring to?

**Comment by Gaztin** [comment]
> 
> 
> > I think we should keep getters as one-liners. The header files look a little busy to me!
> 
> Which getters are you referring to?

Few exampels:
* https://github.com/TheChernoCommunity/GameProject-1/pull/37/files#diff-503777fc6f0303fcabe36e34bb1c8cb8805ae3430cab55dc9cc6ffa4b4c84160R34-R52
* https://github.com/TheChernoCommunity/GameProject-1/pull/37/files#diff-503777fc6f0303fcabe36e34bb1c8cb8805ae3430cab55dc9cc6ffa4b4c84160R64-R77
* https://github.com/TheChernoCommunity/GameProject-1/pull/37/files#diff-503777fc6f0303fcabe36e34bb1c8cb8805ae3430cab55dc9cc6ffa4b4c84160R88-R101

Seems to be all the event types

**Comment by Gaztin** [comment]
[Texture Support](https://trello.com/c/eLCW0qPy/24-texture-support)

**Pull request merged**